networkoperator/v1/types_egressrouter.go

```diff
@@ -0,0 +1,274 @@
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EgressRouter is a feature allowing the user to define an egress router
+// that acts as a bridge between pods and external systems. The egress router runs
+// a service that redirects egress traffic originating from a pod or a group of
+// pods to a remote external system or multiple destinations as per configuration.
+//
+// It is consumed by the cluster-network-operator.
+// More specifically, given an EgressRouter CR with <name>, the CNO will create and manage:
+// - A service called <name>
+// - An egress pod called <name>
+// - A NAD called <name>
+//
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+//
+// EgressRouter is a single egressrouter pod configuration object.
+// +k8s:openapi-gen=true
+// +kubebuilder:resource:path=egressrouters,scope=Namespaced
+// +kubebuilder:printcolumn:name="Condition",type=string,JSONPath=".status.conditions[*].type"
+// +kubebuilder:printcolumn:name="Status",type=string,JSONPath=".status.conditions[*].status"
+type EgressRouter struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Specification of the desired egress router.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinProperties:=4
+	// +kubebuilder:validation:MaxProperties:=4
```

Comment:
these two don't belong here. Make things required (as done here with the spec)

Reply:
hmm, I can make them required. The scenario is: We have 3 modes for egress router. Right now we support only Redirect mode. In future we _may_ support DNS and http modes.

So then other optional fields such as DNSConfig and HTTPConfig will get added. Depending on the mode value, exactly one out of these 3 configs should be passed (somewhat like https://github.com/openshift/api/blob/8494fe80b915542ea013e711edab1365bdaa07bc/operator/v1/types_network.go#L96).

Hence we cannot make those 3 fields (redirect/http/dns) required but we still need a way to check if user has provided at least one out of these 3 config's. However doing
`minProperties` and `maxProperties` though it seems to be supported here: https://master.book.kubebuilder.io/reference/markers/crd-validation.html still doesn't get generated in the yaml for this api. Not sure why.

Any suggestions on how we can move forward if I remove these two checks ?

Reply:
reference: https://coreos.slack.com/archives/CDCP2LA9L/p1613654960279300

Reply:
min/maxProperties don't do what you want. They restrict the number of fields.

What you are after is a union type with a discriminator (mode), compare the (not fully implemented!) KEP https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/1027-api-unions/README.md.

Reply:
yea, we were (mis) using the properties restriction validator to achieve this. If this is not recommended then maybe we can just document this. cc @danwinship

Reply:
If you really want to enforce fields to empty or unset, there are ways with OpenAPI (structural value-) validation:

```
properties:
  mode:
    type: string
  DNSConfig:
    type: string
anyOf:
- properties:
    mode:
      enum:
      - Redirect
    DNSConfig:
      enum:
      - ""
- properties:
    mode:
      enum:
      - DNS
    DNSConfig:
      minLength: 1
```

or similar.

Reply:
aha ok let me try this and see if we can have the redirect mode this way.